### PR TITLE
Improved `callback_context`

### PR DESCRIPTION
**Pull request by AnnMarieW**
closes # https://github.com/plotly/dash/issues/1818
https://github.com/plotly/dash/issues/1054

See also: [Community Forum Discussion](https://community.plotly.com/t/dash-2-x-feature-preview-simplified-callbacks/61738/7)

#### This PR adds features to `dash.callback_context`  to make it easier to use:
-  simplifies the syntax for determining which input triggered a callback. 
- adds `ctx` as an alias for `dash.callback_context`
- adds ability access the ctx dictionary with the dot notation
- simplifies the callback function signature when there are many inputs. 
- makes the dict id easier to access in callbacks using pattern matching. 

#### Determining which input triggered a callback
 The new features will replace boilerplate  such as:

```python
button_id = dash.callback_context.triggered[0]['prop_id'].split('.')[0]
```

Currently, when you have multiple Inputs in a callback, you  might do something like:

```python
@callback(
    Output("graph", "figure"), Input("btn-1", "n_clicks), Input("btn-2", "n_clicks)
)
def display(btn1, btn2):
    ctx = dash.callback_context
    button_id = ctx.triggered[0]['prop_id'].split('.')[0]

    if button_id = "btn-1":
         return update_graph1()
    if button_id = "btn-2":
         return update_graph2()
```

Here one new way introduced by this PR

```python
@callback(
    Output("graph", "figure"), Input("btn-1", "n_clicks), Input("btn-2", "n_clicks)
)
def display(btn1, btn2):
   if  "btn-1.n_clicks" in ctx.triggered_ids:
       return update_graph1()
   if  "btn-2.n_clicks" in ctx.triggered_ids:
      return update_graph2()
```

#### :point_right:  New:  `dash.ctx`  

ctx is a commonly used alias  for  `dash.callback_context`.  The shorter name makes it convenient to add with the import statements, for example:
```
from dash import Dash, html, Input, Output, ctx
```

#### :point_right: New:    `ctx.triggered_ids` 
`triggered_ids` is a `CallbackContext` method that returns a dict of ids that triggered the callback.    When using pattern matching callbacks, it also has the dictionary form of the ids rather than just the stringified version that's  in `callback_context.triggered`.

- keys  (str) : the triggered "prop_id",  
 - values (str or dict):  the id of the component that triggered the callback.          

For example, with regular callback, the ctx.triggered_ids dict would look something like:
```python
{"btn-1.n_clicks": "btn-1"}
```

With  pattern matching ids:
```python
{'{"index":0,"type":"filter-dropdown"}.value': {"index":0,"type":"filter-dropdown"}}
```

----------------------
---------------------

## Update 

Here is a summary of the discussion with  @alexcjohnson and @chriddyp on the items shown above:

`ctx.triggered_ids` is an improvement over the current syntax, but will still be hard to use with long or complicated ids, such as pattern matching, or auto-generated ids.

A better way would be something like:

```python
if variable_name.triggered:
     do_something()
```

We discussed having a type of input called `Triggered`, but that adds complexity to the callback and is not beginner friendly.

I discovered that by using [flexible callback signatures](https://dash.plotly.com/flexible-callback-signatures),  it's not necessary to make any changes to the callback dependencies.

### :tada: Version II:

```python
@app.callback(
    Output("container", "children"),
    inputs=dict(btn1=Input("btn-1", "n_clicks"), btn2=Input("btn-2", "n_clicks")),
)
def display(btn1, btn2):
    c = ctx.args_grouping
    if c.btn1.triggered:
        return f"Button 1 clicked {btn1} times"
    elif c.btn2.triggered:
        return f"Button 2 clicked {btn2} times"
    else:
        return "No clicks yet"
```

#### :point_right: New Improved:  `ctx.args_grouping`

`args_grouping` is a `CallbackContext` method that returns a dict of the inputs when using flexible callback signatures.

- "id": (string or dict) the component id.  If it's a pattern matching id, it will be a dict.
- "id_str": (str) for pattern matching ids, it's the strigified dict id with no white spaces.
- "property": (str) The component property
- "value": the value of the component property at the time the callback was fired
- "triggered":  (bool)Whether this input triggered the callback

So in the above example, if you click the Button 1, then `ctx.args_grouping` looks like:

```
{
    'btn1': {'id': 'btn-1', 'property': 'n_clicks', 'triggered': True, 'value': 1},
    'btn2': {'id': 'btn-2',   'property': 'n_clicks', 'triggered': False,  'value': None}
}
```

#### :point_right: New: Access the `ctx.args_grouping` dict with a dot notation:

```python
c = ctx.args_grouping

# you can access values in the dict like this:
c["btn1"]["value"]

# Or you can use the dot notation:
c.btn1.value

# See if it triggered the callback like this:
if c.btn1.triggerd:
   do_something()
```

----------------

#### Bonus 1:  Simplified callback function signature
See issue #1810 [ Feature Request: Support Dash callback closer to function signature](https://github.com/plotly/dash/issues/1810)

Here's the other great thing about the new `ctx.args_grouping`:  You don't need to repeat all the variable names in the callback function.  This is super convenient  when you have a large callback with many inputs.

##### :cake: In the example below, note that  `all_inputs` is the only arg in the callback function:

```python

@app.callback(
    Output("container", "children"),
    inputs={
        "all_inputs": {
            "btn1": Input("btn-1", "n_clicks"),
            "btn2": Input("btn-2", "n_clicks"),
            "dropdown": Input("dropdown", "value"),
            "store": Store("store", "data"),
        }
    },
)
def display(all_inputs):
    c = ctx.args_grouping.all_inputs  

    if c.btn1.triggered:
        return f"Button 1 clicked {c.btn1.value} times"
    elif c.btn2.triggered:
        return f"Button 2 clicked {c.btn2.value} times"
   #  ... etc
    else:
        "no update"


```

#### Bonus 2:  Easier to get the dict id when using pattern matching callbacks.

It's often necessary to get the component's dictionary id in a callback.  However, the current `dash.callback_context.triggered`. only has the  `"prop_id"` which is a stringified dictionary with no white space. It's necessary to parse the string and use json loads to get the dictionary version.

Here is an example of the current `dash.callback_context.triggered`.
```
[
  {
    'prop_id': '{"index":0,"type":"filter-dropdown"}.value',
    'value': 'NYC'
  }
]
```

__Here is an example app using the new `ctx.triggered_ids` to get the index of the button that triggered the callback__
```python
import numpy as np
import plotly.express as px
from dash import Dash, Input, Output, ctx, dcc, html, ALL

N=5

app = Dash(__name__)


def make_figure(n):
    x = np.linspace(0, 8, 81)
    y = x ** int(n)
    return px.scatter(x=x, y=y)


app.layout = html.Div(
    [
        html.Div(
            [html.Button(f"x^{i}", id={"index":i}) for i in range(N)]
        ),
        dcc.Graph(id="graph"),
    ]
)

@app.callback(
    Output("graph", "figure"), Input({"index":ALL}, "n_clicks")
)
def update_graph(btns):
    if ctx.triggered:
        # The ctx.triggered_ids is a dict with the ids of the input that triggered the callback
        # This gets the key, which is the stringified dict id: "  
        prop_id=ctx.triggered_ids.first()
        
        # This gets the value which is the component's dictionary `id`:
        dict_id= ctx.triggered_ids[prop_id]
        
        # This gets the index number of the button the triggered the callback.  Note you can 
        # use the dot notation or dict_id["index"]
        n= dict_id.index
        
        return make_figure(n)
    return {}


if __name__ == "__main__":
    app.run_server(debug=True)

```

---
__Here is the same app, but instead it uses the new `ctx.args_grouping` to get the index of the button that triggered the callback:__

```python
import numpy as np
import plotly.express as px
from dash import Dash, Input, Output, ctx, dcc, html, ALL

N=5

app = Dash(__name__)


def make_figure(n):
    x = np.linspace(0, 8, 81)
    y = x ** int(n)
    return px.scatter(x=x, y=y)


app.layout = html.Div(
    [
        html.Div(
            [html.Button(f"x^{i}", id={"index":i}) for i in range(N)]
        ),
        dcc.Graph(id="graph"),
    ]
)

@app.callback(
    Output("graph", "figure"), dict(btns=Input({"index":ALL}, "n_clicks"))
)
def update_graph(btns):
    # Since we are using {"index":ALL}, then
    # ctx.args_grouping.btns is a list of dicts - one dict for each button.
    for btn in ctx.args_grouping.btns:
        if btn.triggered:
            n = btn.id.index
            return make_figure(n)
    return {}


if __name__ == "__main__":
    app.run_server(debug=True)

```

### :tada: Version III:

Based on community feedback, 
- renamed `ctx.triggered_ids` to `ctx.triggered_prop_ids`
- added `ctx.triggered_id`, which is the component id of the the Input that triggered the callback.

So now you can use either:
```
if "btn-1" == ctx.triggered_id
   do_something()
```
or
```
if "table-1.selected_rows" in ctx.triggered_prop_ids
    do_something()
if "table-1.selected_columns" in ctx.triggered_prop_ids
    do_something_else()
```

And now, the pattern matching example above can be simplified to:
```
btn = ctx.triggered_id
n = btn.id.index
return make_figure(n)
```

### CODE REVIEW

File: dash/__init__.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from . import html  # noqa: F401,E402
 from . import dash_table  # noqa: F401,E402
 from .version import __version__  # noqa: F401,E402
-from ._callback_context import callback_context  # noqa: F401,E402
+from ._callback_context import callback_context, ctx  # noqa: F401,E402
```
Comment (alexcjohnson) [inline dash/__init__.py:22]: Does this need to be a separate instance? Can we just say `ctx = callback_context` after all the imports, and remove the [extra instantiation](https://github.com/plotly/dash/pull/1952/files#diff-f19c52717a102d5ef8b91b157d234b9075d07bbcc2cd1d0a0cd8060bc9c7f608R290)?

**Comment by alexcjohnson** [comment]
As just discussed on slack, this looks great! Aside from my comment above, this just needs:
- [ ] a changelog entry
- [ ] resolving conflicts
- [ ] flattening the docstring indentation [here](https://github.com/plotly/dash/pull/1952/files#diff-f19c52717a102d5ef8b91b157d234b9075d07bbcc2cd1d0a0cd8060bc9c7f608R54-R57) and [here](https://github.com/plotly/dash/pull/1952/files#diff-f19c52717a102d5ef8b91b157d234b9075d07bbcc2cd1d0a0cd8060bc9c7f608R75-R78)

**Comment by AnnMarieW** [comment]
@alexcjohnson I completed the following:  

> As just discussed on slack, this looks great! Aside from my comment above, this just needs:
> 
> * [x]  a changelog entry
> * [x]  resolving conflicts
> * [x]  flattening the docstring indentation [here](https://github.com/plotly/dash/pull/1952/files#diff-f19c52717a102d5ef8b91b157d234b9075d07bbcc2cd1d0a0cd8060bc9c7f608R54-R57) and [here](https://github.com/plotly/dash/pull/1952/files#diff-f19c52717a102d5ef8b91b157d234b9075d07bbcc2cd1d0a0cd8060bc9c7f608R75-R78)

Also, I removed the extra `callback_context` instance per your comment.
One test  in `test-39` fails, but it's unrelated to any changes in this PR --  I'm not sure what's wrong.

**Review by alexcjohnson — APPROVED** [review]
Fantastic! 💃

**Comment by MattesR** [comment]
IDK whether this is the right spot. I'd just want to mention that there's a typo in this example:
example app using the new ctx.triggered_ids to get the index of the button that triggered the callback:

it says `prop_id=ctx.triggered_ids.first()` but it should be  `prop_id=ctx.triggered_id.first()`. Found out when trying to update my code with the new ctx.

`ctx` has the attributes `'triggered_id'`, and  `'triggered_prop_ids'` according to `dir(ctx)`

**Comment by AnnMarieW** [comment]
@MattesR Thanks for your comment.  I think the examples in the initial write-up for this PR were confusing because the names changed during review.  For example  `triggered_ids` was renamed to `triggered_prop_ids`.   Then a new attribute was added: `ctx.triggered_id`, which is the component id of the the Input that triggered the callback.

I just updated the comment to remove all the examples here and point people to the ones in the announcement on the community forum.

**Pull request merged**